### PR TITLE
Introduce a special DeclBaseName for "deinit"

### PR DESCRIPTION
**Pull request by ahoppen**
## Problem statement
Currently, destructors are represented by the known identifier "deinit", making it impossible to have a function or instance variable named "deinit" in a class. Declaring a function named "deinit" currently results in an error about an invalid redeclaration of `deinit` and adding an instance variable named "deinit" crashes the compiler.

## Solution
Similar to how special `DeclName`s were introduced for subscripts to avoid the name collision between functions named "subscript" and proper subscripts, introduce a special `DeclName` for destructors and remove "deinit" from the list of known identifiers, making the identifier named "deinit" and the name representing the destructor completely distinct and thus resolving the above issue.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -420,7 +420,7 @@ bool Decl::isPrivateStdlibDecl(bool whitelistProtocols) const {
   if (auto AFD = dyn_cast<AbstractFunctionDecl>(D)) {
     // Hide '~>' functions (but show the operator, because it defines
     // precedence).
-    if (AFD->getNameStr() == "~>")
```
Comment (ahoppen) [inline lib/AST/Decl.cpp]: The only reference to the `~>` operator I could find was in Policy.swift which stated that it is a workaround for rdar://problem/14011860. Is that still up to date? GenericDispatch.swift seems to test it but it's used nowhere in the standard library.

File: lib/SIL/SILPrinter.cpp [hunk 2]
```diff
@@ -239,7 +239,8 @@ static void printValueDecl(ValueDecl *Decl, raw_ostream &OS) {
 
   if (Decl->isOperator()) {
     OS << '"' << Decl->getBaseName() << '"';
-  } else if (Decl->getBaseName() == "subscript") {
+  } else if (Decl->getBaseName() == "subscript" ||
+             Decl->getBaseName() == "deinit") {
```
Comment (ahoppen) [inline lib/SIL/SILPrinter.cpp]: I'm not quite sure if we should escape `deinit` in SIL if used as a normal identifier. For subscripts we needed to, since otherwise the names of subscript getters/setters could have clashed with the getters/setters of instance variables named "subscript". AFAIK destructors in SIL always have the suffix `!destructor` so there's no name clash. Should we strive for consistency between the two keywords or simplicity on `deinit`?

**Comment by ahoppen** [comment]
@swift-ci Please test

File: lib/AST/Decl.cpp (hunk 1 above)
Comment (jrose-apple) [inline lib/AST/Decl.cpp]: Aha, yes, that's gone now that we have protocol extensions. Still, that should probably be a separate PR.

File: lib/SIL/SILPrinter.cpp (hunk 2 above)
Comment (jrose-apple) [inline lib/SIL/SILPrinter.cpp]: I'm still inclined to escape all keywords, always, but cc @swiftix, @jckarter.

File: lib/AST/ASTMangler.cpp [hunk 3]
```diff
@@ -548,6 +548,9 @@ void ASTMangler::appendDeclName(const ValueDecl *decl) {
     case DeclBaseName::Kind::Subscript:
       appendIdentifier("subscript");
       break;
+    case DeclBaseName::Kind::Destructor:
+      // FIXME: Is it true that this is unreachable?
```
Comment (jrose-apple) [inline lib/AST/ASTMangler.cpp]: I'm 90% sure this is true but try defining a local type inside a destructor to make sure.

File: include/swift/AST/Decl.h [hunk 4]
```diff
@@ -5647,6 +5647,8 @@ class ConstructorDecl : public AbstractFunctionDecl {
                   GenericParamList *GenericParams, 
                   DeclContext *Parent);
 
+  Identifier getName() const { return getFullName().getBaseIdentifier(); }
```
Comment (jrose-apple) [inline include/swift/AST/Decl.h]: Do we actually use this one? It's always going to be `Id_init`.

**Review by jrose-apple — COMMENTED** [review]
Wow, this fell out very nicely. Good work, Alex.

File: lib/AST/Decl.cpp (hunk 1 above)
Comment (ahoppen) [inline lib/AST/Decl.cpp]: OK, I'll have a look how much of it can be removed later on.

File: include/swift/AST/Decl.h (hunk 4 above)
Comment (ahoppen) [inline include/swift/AST/Decl.h]: I left it in for consistency. You can call `getName()` on all Decls that are guaranteed to be backed by Identifiers. But I don't think it's actually used. Long term goal would be to remove it once/if `Id_init` becomes a special name as well.

File: lib/AST/ASTMangler.cpp (hunk 3 above)
Comment (ahoppen) [inline lib/AST/ASTMangler.cpp]: OK, thanks for the reassurance.

**Comment by dabrahams** [comment]
The way the PR description is phrased this sounds like some kind of troll.  Just sayin' .

**Comment by ahoppen** [comment]
@dabrahams That's certainly not what I wanted. Thanks for the feedback. I've updated it and hope it sounds more serious now.

File: lib/SIL/SILPrinter.cpp (hunk 2 above)
Comment (ahoppen) [inline lib/SIL/SILPrinter.cpp]: @swiftix @jckarter Any opinion on whether "deinit" should be escaped or not?

Comment (jrose-apple) [inline lib/SIL/SILPrinter.cpp]: Ah, I forgot @swiftix was out on vacation. He'll be back tomorrow.

Comment (jckarter) [inline lib/SIL/SILPrinter.cpp]: I agree with Jordan. Escaping all keywords seems like the right conservative answer.

**Comment by ahoppen** [comment]
@jrose-apple Could you have another look through this? I think all the questions have been answered and your comments taken care of.

@swift-ci Please test

**Review by jrose-apple — APPROVED** [review]
We left off with the question about escaping in SIL, and Joe agreed with me that it would be better to escape *everything.* But that doesn't have to be part of this patch, which is just about `deinit`. Yes, this looks good now!

**Comment by ahoppen** [comment]
Ah, I see, I thought by all keywords you meant escaping all keywords that clashed with special decl names. I'll create another PR for it then.

@swift-ci Please test and merge

**Comment by ahoppen** [comment]
Test once again before merging since old test is already several days old.
Btw: Any idea why swift-ci wasn't able to merge the PR?

@swift-ci Please smoke test and merge

**Comment by ahoppen** [comment]
swift-ci didn't run. Let's try again

@swift-ci Please smoke test and merge
